# <center>Shri Ramdeobaba College of Engineering and Management, Nagpur</center>
## <center>Department of Computer Science and Engineering</center>
## <center>Natural Language Processing Lab</center>

**Name** : Shantanu Mane
**Branch** : CSE - AIML (VI<sup>th</sup> SEM)
**Roll Num** : E-63

<h2> AIM :</h2>

1. Write a Python NLTK program to split the text sentence/paragraph from genesis corpus and display it into a list of words. Remove the Punctuation and Stopwords from the given text and perform Stemming and POS tagging.
2. Write a Python program to tokenize sentences with nltk, spacy and gensim in  language other than English (german).

## 1. Importing the Dependencies

In [1]:
from nltk.corpus import genesis
from nltk.corpus import stopwords
from nltk.tokenize import RegexpTokenizer
from nltk.stem import PorterStemmer
import spacy

## Part A
## 2. Working with the Genesis Corpus

### 2.1 Analyzing the Genesis Corpus

In [2]:
genesis.fileids()

['english-kjv.txt',
 'english-web.txt',
 'finnish.txt',
 'french.txt',
 'german.txt',
 'lolcat.txt',
 'portuguese.txt',
 'swedish.txt']

In [3]:
genesis.words('english-web.txt')

['In', 'the', 'beginning', 'God', 'created', 'the', ...]

In [4]:
genesisWords = genesis.words('english-web.txt')

len(genesisWords)

44054

### 2.2 Removing StopWords from the Genesis Corpus Text

#### 2.2.1 List of Predefined Stopwords

In [5]:
stopwordsList = stopwords.words('english')

print(
    f"Some of the stop words : {stopwordsList[:15]}, length of the stopwords list in english language : {len(stopwordsList)}"
)

Some of the stop words : ['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours'], length of the stopwords list in english language : 179


In [6]:
genesisWords_wo_stopwords = [word for word in genesisWords if not word in stopwordsList]

len(genesisWords_wo_stopwords)

25756

As we can see, we have reduced the length of genesis words list from $44054$ to $25756$, which implies that stop words are removed successfully.

In [7]:
genesisWords_wo_stopwords

['In',
 'beginning',
 'God',
 'created',
 'heavens',
 'earth',
 '.',
 'Now',
 'earth',
 'formless',
 'empty',
 '.',
 'Darkness',
 'surface',
 'deep',
 '.',
 'God',
 "'",
 'Spirit',
 'hovering',
 'surface',
 'waters',
 '.',
 'God',
 'said',
 ',',
 '"',
 'Let',
 'light',
 ',"',
 'light',
 '.',
 'God',
 'saw',
 'light',
 ',',
 'saw',
 'good',
 '.',
 'God',
 'divided',
 'light',
 'darkness',
 '.',
 'God',
 'called',
 'light',
 'Day',
 ',',
 'darkness',
 'called',
 'Night',
 '.',
 'There',
 'evening',
 'morning',
 ',',
 'one',
 'day',
 '.',
 'God',
 'said',
 ',',
 '"',
 'Let',
 'expanse',
 'middle',
 'waters',
 ',',
 'let',
 'divide',
 'waters',
 'waters',
 '."',
 'God',
 'made',
 'expanse',
 ',',
 'divided',
 'waters',
 'expanse',
 'waters',
 'expanse',
 ';',
 '.',
 'God',
 'called',
 'expanse',
 'sky',
 '.',
 'There',
 'evening',
 'morning',
 ',',
 'second',
 'day',
 '.',
 'God',
 'said',
 ',',
 '"',
 'Let',
 'waters',
 'sky',
 'gathered',
 'together',
 'one',
 'place',
 ',',
 'let',
 'dr

### 2.3 Removing the Punctuations from Genesis Corpus

#### 2.3.1 Using NLTK's RegExpTokeniser

In [8]:
regexTokenizer = RegexpTokenizer(r'\w+')

In [9]:
' '.join(genesisWords_wo_stopwords)

'In beginning God created heavens earth . Now earth formless empty . Darkness surface deep . God \' Spirit hovering surface waters . God said , " Let light ," light . God saw light , saw good . God divided light darkness . God called light Day , darkness called Night . There evening morning , one day . God said , " Let expanse middle waters , let divide waters waters ." God made expanse , divided waters expanse waters expanse ; . God called expanse sky . There evening morning , second day . God said , " Let waters sky gathered together one place , let dry land appear ;" . God called dry land Earth , gathering together waters called Seas . God saw good . God said , " Let earth put forth grass , herbs yielding seed , fruit trees bearing fruit kind , seed , earth ;" . The earth brought forth grass , herbs yielding seed kind , trees bearing fruit , seed , kind ; God saw good . There evening morning , third day . God said , " Let lights expanse sky divide day night ; let signs , seasons , d

In [10]:
genesisWords_wo_punctuations = regexTokenizer.tokenize(' '.join(genesisWords_wo_stopwords))

genesisWords_wo_punctuations[:50]

['In',
 'beginning',
 'God',
 'created',
 'heavens',
 'earth',
 'Now',
 'earth',
 'formless',
 'empty',
 'Darkness',
 'surface',
 'deep',
 'God',
 'Spirit',
 'hovering',
 'surface',
 'waters',
 'God',
 'said',
 'Let',
 'light',
 'light',
 'God',
 'saw',
 'light',
 'saw',
 'good',
 'God',
 'divided',
 'light',
 'darkness',
 'God',
 'called',
 'light',
 'Day',
 'darkness',
 'called',
 'Night',
 'There',
 'evening',
 'morning',
 'one',
 'day',
 'God',
 'said',
 'Let',
 'expanse',
 'middle',
 'waters']

## 3. Stemming and POS Tagging of Genesis Corpus

### 3.1 Stemming Using `PorterStemmer`

In [11]:
pStemmer = PorterStemmer()

print(f"Word".ljust(15), f"Stem Word")
print("-" * 25)
for word in genesisWords_wo_punctuations[:25]:
    print(f"{word.lower()}".ljust(15), f"{pStemmer.stem(word)}")

Word            Stem Word
-------------------------
in              in
beginning       begin
god             god
created         creat
heavens         heaven
earth           earth
now             now
earth           earth
formless        formless
empty           empti
darkness        dark
surface         surfac
deep            deep
god             god
spirit          spirit
hovering        hover
surface         surfac
waters          water
god             god
said            said
let             let
light           light
light           light
god             god
saw             saw


### 3.2 Part Of Speech Tagging Using SpaCy

In [12]:
spacyObj = spacy.load('en_core_web_sm')

In [13]:
genesisWords_Doc = spacyObj(" ".join(genesisWords_wo_punctuations))

genesisWords_Doc[:100]

In beginning God created heavens earth Now earth formless empty Darkness surface deep God Spirit hovering surface waters God said Let light light God saw light saw good God divided light darkness God called light Day darkness called Night There evening morning one day God said Let expanse middle waters let divide waters waters God made expanse divided waters expanse waters expanse God called expanse sky There evening morning second day God said Let waters sky gathered together one place let dry land appear God called dry land Earth gathering together waters called Seas God saw good God said Let

In [14]:
print(f"Word".ljust(15), "Alpha", "Space", "Stop", "Punctuation")
print("-" * 47)
for word in genesisWords_Doc[:25]:
    print(f"{word}".ljust(15), word.is_alpha, '', word.is_space, word.is_stop, word.is_punct)

Word            Alpha Space Stop Punctuation
-----------------------------------------------
In              True  False True False
beginning       True  False False False
God             True  False False False
created         True  False False False
heavens         True  False False False
earth           True  False False False
Now             True  False True False
earth           True  False False False
formless        True  False False False
empty           True  False True False
Darkness        True  False False False
surface         True  False False False
deep            True  False False False
God             True  False False False
Spirit          True  False False False
hovering        True  False False False
surface         True  False False False
waters          True  False False False
God             True  False False False
said            True  False False False
Let             True  False False False
light           True  False False False
light           True  False Fa

## Part B
## 4. Tokenizing German Language Using NLTK

### 4.1 Analyzing the German language from Genesis Corpus

In [15]:
# German language corpus is inbuilt in Genesis Corpus

genesis.fileids()[4]

'german.txt'

In [16]:
germanTokens = genesis.words('german.txt')

germanTokens

['Am', 'Anfang', 'schuf', 'Gott', 'Himmel', 'und', ...]

In [17]:
len(germanTokens)

43941

### 4.2 Removing stopwords from text

In [18]:
germanStopWords = set(stopwords.words('german'))

germanTokens[:10]

['Am', 'Anfang', 'schuf', 'Gott', 'Himmel', 'und', 'Erde', '.', 'Und', 'die']

In [19]:
germanTokens_wo_stopwords = [word for word in germanTokens if not word in germanStopWords]

print(f"text without stopwords : {germanTokens_wo_stopwords[:10]}, length of text : {len(germanTokens_wo_stopwords)}")

text without stopwords : ['Am', 'Anfang', 'schuf', 'Gott', 'Himmel', 'Erde', '.', 'Und', 'Erde', 'wüst'], length of text : 26699


As we can see, we have reduced the length of genesis words list from $43941$ to $26699$, which implies that stop words are removed successfully.

### 4.3 Removing punctuations using `RegexpTokenizer`

In [20]:
punctTokenizer = RegexpTokenizer(r'\w+')

germanWords_wo_punct = punctTokenizer.tokenize(" ".join(germanTokens_wo_stopwords))

print(f"text without punctuations : {germanWords_wo_punct[:10]}, length of text : {len(germanWords_wo_punct)}")

text without punctuations : ['Am', 'Anfang', 'schuf', 'Gott', 'Himmel', 'Erde', 'Und', 'Erde', 'wüst', 'leer'], length of text : 18831


### 4.4 POS Tagging using SpaCy

In [23]:
spacyObjGerman = spacy.load("de_core_news_sm")

germanWords_Doc = spacyObjGerman(" ".join(germanTokens))

germanWords_Doc[:10]

Am Anfang schuf Gott Himmel und Erde . Und die

In [24]:
print(f"Word".ljust(15), "Alpha", "Space", "Stop", "Punctuation")
print("-" * 47)
for word in germanWords_Doc[:25]:
    print(f"{word}".ljust(15), word.is_alpha, '', word.is_space, word.is_stop, word.is_punct)

Word            Alpha Space Stop Punctuation
-----------------------------------------------
Am              True  False True False
Anfang          True  False False False
schuf           True  False False False
Gott            True  False False False
Himmel          True  False False False
und             True  False True False
Erde            True  False False False
.               False  False False True
Und             True  False True False
die             True  False True False
Erde            True  False False False
war             True  False True False
wüst            True  False False False
und             True  False True False
leer            True  False False False
,               False  False False True
und             True  False True False
es              True  False True False
war             True  False True False
finster         True  False False False
auf             True  False True False
der             True  False True False
Tiefe           True  False False Fals